src/frontend/flight/sync/schema.rs

```diff
@@ -0,0 +1,128 @@
+                        )
+                    } else if field.is_nullable() {
+                        format!(
+                            "Field for column with `Changed` role can not be nullable: {}",
+                            field.name()
+                        )
+                    } else if !column_descriptors.iter().any(|other_cd| {
+                        col_desc.name == other_cd.name
+                            && other_cd.role == ColumnRole::Value as i32
+                    }) {
+                        format!(
+                            "Column with `Changed` role must point to an existing column with `Value` role: {}",
+                            &col_desc.name
+                        )
+                    } else {
+                        // All good
+                        continue;
+                    };
+                    warn!(err);
+                    return Err(SyncError::SchemaError { reason: err });
+                }
+                Err(err) => {
+                    let err = format!("Failed parsing role: {err:?}");
+                    warn!(err);
+                    return Err(SyncError::SchemaError { reason: err });
+                }
+            }
+        }
+        if old_pk_count == 0 || new_pk_count == 0 || old_pk_count != new_pk_count {
+            let err = "Change requested but batches do not contain old/new PK columns";
+            warn!(err);
+            return Err(SyncError::SchemaError {
+                reason: err.to_string(),
+            });
+        }
+
+        Ok(Self {
+            column_descriptors,
+            schema,
+        })
+    }
+
+    #[allow(dead_code)]
+    pub fn arrow_schema(&self) -> SchemaRef {
+        self.schema.clone()
+    }
+
+    pub fn column(&self, name: &str, role: ColumnRole) -> Option<SyncColumn> {
+        self.columns()
+            .find(|col| col.name == name && col.role == role)
+    }
+
+    pub fn columns(&self) -> impl Iterator<Item = SyncColumn> + '_ {
+        self.column_descriptors
+            .iter()
+            .zip(self.schema.fields())
+            .map(|(column_descriptor, field)| SyncColumn {
+                role: column_descriptor.role(),
+                name: column_descriptor.name.clone(),
+                field: field.clone(),
+            })
+    }
+}
+
+pub struct SyncColumn {
+    role: ColumnRole,
+    name: String,
+    field: FieldRef,
+}
+
+impl SyncColumn {
+    // Get the role of the column
+    pub fn role(&self) -> ColumnRole {
+        self.role
+    }
+
+    // Name of the column that this descriptor refers to. Note that in general this differs from the
+    // underlying arrow field name. You can think of them as physical and logical names, respectively.
+    //
+    // For all roles except `Changed`, the sync column refers to itself, but for `Changed` it refers
+    // to a `Value` column whose field name is the same as this name.
+    pub fn name(&self) -> &String {
+        &self.name
```

Comment:
How are you allowed to return a reference without a lifetime?

Reply:
It's an implicit reference, I think the compiler inferred it from `&self`

Reply:
Shows how little I know :)